### PR TITLE
Support jsonMerge function

### LINKED ISSUE #54102: Support jsonMerge function
**Use case**

```sql
select jsonMerge(json_doc, json_doc [, json_doc] ...)
```

returns a merged JSON object

**Additional context**

See also https://github.com/ClickHouse/ClickHouse/issues/45068.

### PR DESCRIPTION
**Pull request by Joeywzr**
Close https://github.com/ClickHouse/ClickHouse/issues/54102
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Added function `jsonMergePatch`. When working with JSON data as strings, it provides a way to merge these strings (of JSON objects) together to form a single string containing a single JSON object.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

- Motivation: When working with JSON data as strings, provide a way to merge these strings (of JSON objects) together to form a single string containing a single JSON object.

- Parameters: JSON Objects like '{"a":1}', '{"a": "1","b": 2,"c": [true,{"qrdzkzjvnos": true,"yxqhipj": false,"oesax": "33o8_6AyUy"}]}', '{"c": "1"}'

- Example use: 

input:
select jsonMergePatch('{"a":1}','{"name": "joey"}','{"name": "tom"}','{"name": "zoey"}');

output:
┌────────────┐
  │ {"a":1,"name":"zoey"}         │
└────────────┘

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll2** [comment]
This is an automated comment for commit 69fd9c8d57b22ba4ada9c3b5cae213c35ab9ba8b with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54364/69fd9c8d57b22ba4ada9c3b5cae213c35ab9ba8b/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🔴 failure</td></tr>
</table>

File: src/Functions/jsonMerge.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+        DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+        {
+            if (arguments.empty())
+                throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {} requires at least one argument.", getName());
+
+            return std::make_shared<DataTypeString>();
+        }
+
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
```
Comment (Joeywzr) [inline src/Functions/jsonMerge.cpp:61]: I'm not sure about what `input_rows_count` means. Can someone help me explain it?

**Review by CurtizJ — COMMENTED** [review]
Well, I took MySQL for the reference. It has two functions: [JSON_MERGE_PATCH](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-merge-patch) and [JSON_MERGE_PRESERVE](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-merge-preserve). The first one replaces conflicting keys, the second one combines them into an array. 

You implemented a kind of `JSON_MERGE_PATCH` analog, but `JSON_MERGE_PATCH` merges nested paths as well:

```sql
SELECT JSON_MERGE_PATCH('{"a": {"b": 1, "c": 2}}', '{"a": {"b": [3, 4]}}');

{"a": {"b": [3, 4], "c": 2}}
```

I find this behaviour reasonable, so please update your implementation to support merging of nested paths.

File: src/Functions/jsonMerge.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                    {
+                        rapidjson::Value key(it->name, allocator);
+                        rapidjson::Value value(it->value, allocator);
+                        if (merged_json.HasMember(key))
+                            merged_json[key] = value;
+                        else
+                            merged_json.AddMember(key, value, allocator);
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (Joeywzr) [reply src/Functions/jsonMerge.cpp:92]: You are right. I missed this case. Working on it.

**Comment by Joeywzr** [comment]
@CurtizJ Please take a look.

**Review by CurtizJ — APPROVED** [review]

**Comment by pauldemarco** [comment]
This is excellent. Is there somewhere to track support of [JSON_MERGE_PRESERVE](https://dev.mysql.com/doc/refman/8.0/en/json-modification-functions.html#function_json-merge-preserve)?

**Comment by alexey-milovidov** [comment]
@CurtizJ, json is an abbreviation, and should be `JSON`, like in other function names.

**Pull request merged**